features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/resources/ResourceDataStoreManager.kt

```diff
@@ -0,0 +1,158 @@
+import com.datadog.android.core.persistence.Serializer
+import com.datadog.android.core.persistence.datastore.DataStoreContent
+import com.datadog.android.sessionreplay.internal.ResourcesFeature.Companion.SESSION_REPLAY_RESOURCES_FEATURE_NAME
+import com.datadog.android.sessionreplay.model.ResourceHashesEntry
+import java.util.Collections
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicLong
+
+internal class ResourceDataStoreManager(
+    private val featureSdkCore: FeatureSdkCore,
+    private val resourceHashesSerializer: Serializer<ResourceHashesEntry>,
+    private val resourceHashesDeserializer: Deserializer<String, ResourceHashesEntry>
+) {
+    private val knownResources = Collections.synchronizedSet(mutableSetOf<String>())
+    private var storedLastUpdateDateNs = AtomicLong(System.nanoTime())
+    private var isInitialized = AtomicBoolean(false) // has init finished executing its async actions
+
+    init {
+        val onDeleteDatastoreCallback = object : DataStoreWriteCallback {
+            override fun onSuccess() {
+                finishedInitializingManager()
+            }
+
+            override fun onFailure() {
+                finishedInitializingManager()
+            }
+        }
+
+        fetchStoredResourceHashes(
+            onFetchSuccessful = lambda@{ storedEntry ->
+                val storedData = storedEntry?.data
+
+                if (storedData == null) {
+                    finishedInitializingManager()
+                    return@lambda
+                }
+
+                val lastUpdateDateNs = storedData.lastUpdateDateNs
+                val storedHashes = storedData.resourceHashes
+
+                storedLastUpdateDateNs.set(lastUpdateDateNs.toLong())
+
+                if (didDataStoreExpire()) {
+                    handleDataStoreExpired(
+                        storedHashes = storedHashes.toSet(),
+                        callback = onDeleteDatastoreCallback
+                    )
+                } else {
+                    knownResources.addAll(storedHashes)
+                    finishedInitializingManager()
+                }
+            },
+            onFetchFailure = {
+                deleteStoredHashesEntry(onDeleteDatastoreCallback)
+            }
+        )
+    }
+
+    internal fun isPreviouslySentResource(resourceHash: String): Boolean =
+        knownResources.contains(resourceHash)
+
+    internal fun cacheResourceHash(resourceHash: String) {
+        knownResources.add(resourceHash)
+        writeResourcesToStore()
+    }
+
+    internal fun isReady(): Boolean =
+        isInitialized.get()
+
+    // region internal
+
+    private fun finishedInitializingManager() {
+        isInitialized.set(true)
+    }
+
+    private fun writeResourcesToStore() {
+        val data = ResourceHashesEntry(
+            lastUpdateDateNs = storedLastUpdateDateNs,
+            resourceHashes = knownResources.toList()
+        )
```

Comment:
probably not the most efficient design. Imagine you have 1000 hashes, each is 36 characters (UUID length). All chars are ASCII, so each char is 1 byte. It means that once you add 1001th hash, you need to write (and also serialize) all 1000 hashes you wrote before, which is 36KB. So with each write operation you have data overhead of 36KB, instead of storing only the hash you need to.

Reply:
as discussed, we'll leave this for now and perhaps improve in a future pr